rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/selection/SelectionManager.java

```diff
@@ -262,58 +262,60 @@ private NodeSet doSelectNodes(Criteria criteria, Client client) {
-                        // we can run
-                        // "PAResourceManagerProperties.RM_SELECTION_MAX_THREAD_NUMBER.getValueAsInt()"
-                        // scripts in parallel
-                        // in case when we need less nodes it still useful to
-                        // the full capacity of the thread pool to find nodes
-                        // quicker
-
-                        // it is not important if we find more nodes than needed
-                        // subset will be selected later (topology handlers)
-                        numberOfNodesForScriptExecution = PAResourceManagerProperties.RM_SELECTION_MAX_THREAD_NUMBER.getValueAsInt();
+                if (arrangedFilteredNodes.isEmpty()) {
+                    matchedNodes = new LinkedList<>();
+                } else if (electedToRunOnAllNodes(criteria)) {
+                    // run scripts on all available nodes
+                    matchedNodes = runScripts(arrangedFilteredNodes, criteria);
+                } else {
+                    // run scripts not on all nodes, but always on missing number of
+                    // nodes
+                    // until required node set is found
+                    matchedNodes = new LinkedList<>();
+                    while (matchedNodes.size() < criteria.getSize()) {
+                        int numberOfNodesForScriptExecution = criteria.getSize() - matchedNodes.size();
+
+                        if (numberOfNodesForScriptExecution < PAResourceManagerProperties.RM_SELECTION_MAX_THREAD_NUMBER.getValueAsInt()) {
+                            // we can run
+                            // "PAResourceManagerProperties.RM_SELECTION_MAX_THREAD_NUMBER.getValueAsInt()"
+                            // scripts in parallel
+                            // in case when we need less nodes it still useful to
+                            // the full capacity of the thread pool to find nodes
+                            // quicker
+
+                            // it is not important if we find more nodes than needed
+                            // subset will be selected later (topology handlers)
+                            numberOfNodesForScriptExecution = PAResourceManagerProperties.RM_SELECTION_MAX_THREAD_NUMBER.getValueAsInt();
+                        }
+
+                        List<RMNode> subset = arrangedFilteredNodes.subList(0,
+                                                                            Math.min(numberOfNodesForScriptExecution,
+                                                                                     arrangedFilteredNodes.size()));
+                        matchedNodes.addAll(runScripts(subset, criteria));
+                        // removing subset of arrangedNodes
+                        subset.clear();
+
+                        if (arrangedFilteredNodes.size() == 0) {
```

Comment:
Why not putting this condition in the while loop ? Usually it is better to avoid using break statement if you can put all conditions in the loop.